File: app/src/main/java/com/werebug/anmapwrapper/ImportNmapAssets.java
package com.werebug.anmapwrapper;

import static com.werebug.anmapwrapper.MainActivity.LOG_TAG;

import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.lang.ref.WeakReference;
import java.util.Objects;

public class ImportNmapAssets implements Runnable {

    private static final String ASSET_VERSION_PREFS_KEY = "last_installed_asset_version";
    private static final String ASSET_VERSION = "7.94";
    private static final String[] NMAP_ASSETS = {"nmap-service-probes",
                                                 "nmap-services",
                                                 "nmap-protocols",
                                                 "nmap-rpc",
                                                 "nmap-mac-prefixes",
                                                 "nmap-os-db"};
    private final WeakReference<MainActivity> mainActivityRef;
    private final SharedPreferences preferences;

    public ImportNmapAssets(WeakReference<MainActivity> mainActivityRef) {
        this.mainActivityRef = mainActivityRef;
        this.preferences = this.mainActivityRef.get().getPreferences(Context.MODE_PRIVATE);
    }

    @Override
    public void run() {
        String lastImportedVersion = preferences.getString(ASSET_VERSION_PREFS_KEY, "");
        for (String assetFileName : NMAP_ASSETS) {
            File assetFile = mainActivityRef.get().getFileStreamPath(assetFileName);
            if (Objects.requireNonNull(assetFile).exists()
                    && lastImportedVersion.equals(ASSET_VERSION)) {
                continue;
            }
            BufferedReader reader = null;
            BufferedWriter writer = null;
            try {
                reader = new BufferedReader(new InputStreamReader(
                        mainActivityRef.get().getAssets().open(assetFileName)));
                writer = new BufferedWriter(new OutputStreamWriter(
                        mainActivityRef.get().openFileOutput(assetFileName, Context.MODE_PRIVATE)));
                String line;
                while ((line = reader.readLine()) != null) {
                    writer.write(line + "\n");
                }
                Log.d(LOG_TAG, assetFileName + " imported.");
            } catch (IOException e) {
                Log.e(LOG_TAG, "Error importing " + assetFileName + ".");
            } finally {
                try {
                    Objects.requireNonNull(reader).close();
                } catch (IOException ignored) {
                }
                try {
                    Objects.requireNonNull(writer).close();
                } catch (IOException ignored) {
                }
            }
        }
        SharedPreferences.Editor editor = preferences.edit();
        editor.putString(ASSET_VERSION_PREFS_KEY, ASSET_VERSION);
        editor.apply();
    }
}

File: app/src/main/java/com/werebug/anmapwrapper/ImportNmapAssets.kt
package com.werebug.anmapwrapper

import android.content.SharedPreferences
import android.content.res.AssetManager
import android.util.Log
import androidx.core.content.edit
import java.io.File
import java.io.FileOutputStream
import java.io.IOException

class ImportNmapAssets(
  private val assets: AssetManager,
  private val filesDir: File,
  private val preferences: SharedPreferences
) : Runnable {

  companion object {
    private const val ASSET_VERSION_PREFS_KEY = "last_installed_asset_version"
    private const val ASSET_VERSION = "7.991"
    private val NMAP_FILE_ASSETS = arrayOf(
      "nmap-service-probes",
      "nmap-services",
      "nmap-protocols",
      "nmap-rpc",
      "nmap-mac-prefixes",
      "nmap-os-db",
      "nse_main.lua"
    )
    private val NMAP_FOLDER_ASSETS = arrayOf("scripts", "nselib")
  }

  override fun run() {
    val lastImportedVersion = preferences.getString(ASSET_VERSION_PREFS_KEY, "")
    var allSucceeded = true
    for (dataAssetFile in NMAP_FILE_ASSETS) {
      if (!copyAssetFileToInternalStorage(dataAssetFile, dataAssetFile, lastImportedVersion)) {
        allSucceeded = false
      }
    }
    for (assetFolder in NMAP_FOLDER_ASSETS) {
      if (!copyAssetDirToInternalStorage(assetFolder, assetFolder, lastImportedVersion)) {
        allSucceeded = false
      }
    }
    if (allSucceeded) {
      preferences.edit {
        putString(ASSET_VERSION_PREFS_KEY, ASSET_VERSION)
      }
    } else {
      Log.e(MainActivity.LOG_TAG, "Asset import incomplete; version pref not updated")
    }
  }

  private fun copyAssetFileToInternalStorage(
    assetPath: String,
    targetPath: String,
    lastImportedVersion: String?
  ): Boolean {
    val targetFile = File(filesDir, targetPath)
    if (targetFile.exists() && lastImportedVersion == ASSET_VERSION) {
      return true
    }
    return try {
      assets.open(assetPath).use { input ->
        FileOutputStream(targetFile).use { output ->
          input.copyTo(output)
        }
      }
      Log.i(MainActivity.LOG_TAG, "$assetPath successfully imported.")
      true
    } catch (e: IOException) {
      Log.e(MainActivity.LOG_TAG, "Error importing $assetPath: ${e.message}")
      false
    }
  }

  private fun copyAssetDirToInternalStorage(
    assetDir: String,
    targetDir: String,
    lastImportedVersion: String?
  ): Boolean {
    val children = assets.list(assetDir)
    if (children == null) {
      Log.e(MainActivity.LOG_TAG, "Error listing asset dir $assetDir")
      return false
    }
    val targetDirectory = File(filesDir, targetDir)
    if (!targetDirectory.exists() && !targetDirectory.mkdirs()) {
      Log.e(MainActivity.LOG_TAG, "Error creating directory ${targetDirectory.path}")
      return false
    }
    var allSucceeded = true
    for (asset in children) {
      val assetPath = if (assetDir.isEmpty()) asset else "$assetDir/$asset"
      val targetPath = "$targetDir/$asset"
      val ok = if (assets.list(assetPath)?.isNotEmpty() == true) {
        copyAssetDirToInternalStorage(assetPath, targetPath, lastImportedVersion)
      } else {
        copyAssetFileToInternalStorage(assetPath, targetPath, lastImportedVersion)
      }
      if (!ok) allSucceeded = false
    }
    return allSucceeded
  }
}
